### PR TITLE
feature: add billing details to create buyer

### PR DESCRIPTION
**Pull request by alibeylan**
# Description

Add billing details to open API.

## Checklist

- [X] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own changes
- [x] I have run `yarn test` to make sure my changes pass all linters
- [x] I have pulled the latest changes from the upstream `main` branch

## Contribution guidelines

For contribution guidelines, styleguide, and other helpful information please
see the `CONTRIBUTING.md` file in the root of this project.

### CODE REVIEW

File: reference/request-bodies/buyer_update_request.yaml [hunk 1]
```diff
@@ -8,3 +8,10 @@ x-tags:
 
 allOf:
   - $ref: "../openapi.yaml#/components/schemas/BuyerRequest"
+
+properties:
+  billing_details:
+    description: The optional billing details to create a buyer.
```
Comment (ipeluffo) [inline reference/request-bodies/buyer_update_request.yaml:14]:
```suggestion
    description: The optional billing details to update a buyer.
```

File: reference/request-bodies/transaction_create_request.yaml [hunk 2]
```diff
@@ -87,3 +87,10 @@ properties:
       - staging
       - production
     example: staging
+
+  buyer_external_identifier:
```
Comment (ipeluffo) [inline reference/request-bodies/transaction_create_request.yaml:91]: `buyer_external_identifier` is a property in the `payment_method` payload, was this changed?

File: reference/resources/address.yaml [hunk 3]
```diff
@@ -0,0 +1,80 @@
+---
+title: Billing details request
```
Comment (ipeluffo) [inline reference/resources/address.yaml:2]:
```suggestion
title: Address
```

File: reference/resources/address.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+type: object
+description: |-
+  Billing details to use associated to a buyer.
```
Comment (ipeluffo) [inline reference/resources/address.yaml:5]:
```suggestion
  Billing details address to associate to a buyer.
```

File: reference/resources/address.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  - state
+  - postal_code
+  - country
+
+properties:
+  city:
+    type: string
+    minLength: 1
+    maxLength: 100
+    description: City to use for this request.
```
Comment (ipeluffo) [inline reference/resources/address.yaml:22]: sorry, maybe I'm a bit confused, why do all fields have "for this request"? Should it be "for this address"?

File: reference/resources/address.yaml [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    example: London
+    nullable: true
```
Comment (ipeluffo) [inline reference/resources/address.yaml:24]: Should this be `false`? (same question for other required properties)

File: reference/request-bodies/transaction_create_request.yaml [hunk 7]
```diff
@@ -87,3 +87,4 @@ properties:
       - staging
       - production
     example: staging
+
```
Comment (cbetta) [inline reference/request-bodies/transaction_create_request.yaml:90]: @alibeylan please run the linter locally. As you can see there's a few errors in the linter.

File: reference/resources/tax_id.yaml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  tax_id:
+    type: string
+    minLength: 1
+    maxLength: 50
+    description: The tax id to use for this request.
+    example: "12345678931"
+
+  tax_id_kind:
+    type: string
+    description: The tax id kind to use for this request.
```
Comment (cbetta) [inline reference/resources/tax_id.yaml:24]:
```suggestion
    description: The unique name for this kind of tax ID.
```

File: reference/resources/tax_id.yaml [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+required:
+  - kind
+  - value
+
+properties:
+  tax_id:
+    type: string
+    minLength: 1
+    maxLength: 50
+    description: The tax id to use for this request.
```
Comment (cbetta) [inline reference/resources/tax_id.yaml:19]:
```suggestion
    description: The tax ID for the buyer.
```

File: reference/resources/tax_id.yaml [hunk 10]
```diff
@@ -0,0 +1,63 @@
+---
+title: Tax ID
+type: object
+description: |-
+  Tax ID to use associated to a buyer.
```
Comment (cbetta) [inline reference/resources/tax_id.yaml:5]:
```suggestion
  The tax ID information associated to a buyer.
```

File: reference/resources/buyer.yaml [hunk 11]
```diff
@@ -38,6 +38,12 @@ properties:
     minLength: 1
     maxLength: 200
 
+  billing_details:
+    description: Billing details associated with the buyer.
```
Comment (cbetta) [inline reference/resources/buyer.yaml:42]:
```suggestion
    description: The billing details associated with a buyer.
```

File: reference/resources/buyer--snapshot.yaml [hunk 12]
```diff
@@ -40,3 +40,10 @@ properties:
     nullable: true
     minLength: 1
     maxLength: 200
+
+  billing_details:
+    description:
+      Billing details, including address and tax id, associated with the buyer.
```
Comment (cbetta) [inline reference/resources/buyer--snapshot.yaml:46]:
```suggestion
      The billing details associated with the buyer, which include the address and tax ID.
```

File: reference/resources/billing_details.yaml [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    description: Email to use for this request.
+    example: john@example.com
+    nullable: true
+
+  phone_number:
+    type: string
+    pattern: '^\+[1-9]\d{1,14}$'
+    minLength: 1
+    maxLength: 50
+    description: Phone number to use for this request following E164 international standard.
```
Comment (cbetta) [inline reference/resources/billing_details.yaml:40]:
```suggestion
    description: |-
      Phone number to use for this request following E164 international standard.
```

File: reference/resources/billing_details.yaml [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    maxLength: 255
+    example: Lunn
+    description: Last name to use for this request.
+    nullable: true
+
+  email:
+    type: string
+    minLength: 1
+    maxLength: 320
+    description: Email to use for this request.
```
Comment (cbetta) [inline reference/resources/billing_details.yaml:31]:
```suggestion
    description: The email address for the buyer
```

File: reference/resources/billing_details.yaml [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    type: string
+    pattern: '^\+[1-9]\d{1,14}$'
+    minLength: 1
+    maxLength: 50
+    description: Phone number to use for this request following E164 international standard.
+    example: "+1234567890"
+    nullable: true
+
+  address:
+    description: Address associated with the billing details.
```
Comment (cbetta) [inline reference/resources/billing_details.yaml:45]:
```suggestion
    description: The billing address for the buyer.
```

File: reference/resources/billing_details.yaml [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    nullable: true
+    allOf:
+      - $ref: "../openapi.yaml#/components/schemas/Address"
+
+  tax_id:
+    description: Tax information associated with the billing details.
```
Comment (cbetta) [inline reference/resources/billing_details.yaml:51]:
```suggestion
    description: The tax information associated with the billing details.
```

File: reference/resources/billing_details.yaml [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    description: First name to use for this request.
+    example: John
+    nullable: true
+
+  last_name:
+    type: string
+    minLength: 1
+    maxLength: 255
+    example: Lunn
+    description: Last name to use for this request.
```
Comment (cbetta) [inline reference/resources/billing_details.yaml:24]:
```suggestion
    description: The last name, or family name, of the buyer.
```

File: reference/resources/billing_details.yaml [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+
+x-tags:
+  - Buyers
+
+properties:
+  first_name:
+    type: string
+    minLength: 1
+    maxLength: 255
+    description: First name to use for this request.
```
Comment (cbetta) [inline reference/resources/billing_details.yaml:15]:
```suggestion
    description: The first name(s) or given name for the buyer.
```

File: reference/resources/billing_details.yaml [hunk 19]
```diff
@@ -0,0 +1,54 @@
+---
+title: Billing details request
+type: object
+description: |-
+  Billing details to use associated to a buyer.
```
Comment (cbetta) [inline reference/resources/billing_details.yaml:5]:
```suggestion
  Billing details associated to a buyer.
```

File: reference/resources/address.yaml [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+    maxLength: 255
+    description: The second line of the address to use for this address.
+    example: "New Oxford Court"
+    nullable: true
+
+  organization:
+    type: string
+    minLength: 1
+    maxLength: 255
+    description: The second line of the address to use for this address.
```
Comment (cbetta) [inline reference/resources/address.yaml:78]:
```suggestion
    description: The optional name of the company or organisation to add to the billing address.
```

File: reference/resources/address.yaml [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+    maxLength: 255
+    description: The house number or name to use for this address.
+    example: "10"
+    nullable: true
+
+  line1:
+    type: string
+    minLength: 1
+    maxLength: 255
+    description: The first line of the address to use for this address.
```
Comment (cbetta) [inline reference/resources/address.yaml:62]:
```suggestion
    description: The first line of the billing address.
```

File: reference/resources/address.yaml [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+    example: "10 Oxford Street"
+    nullable: false
+
+  line2:
+    type: string
+    minLength: 1
+    maxLength: 255
+    description: The second line of the address to use for this address.
```
Comment (cbetta) [inline reference/resources/address.yaml:70]:
```suggestion
    description: The second line of the billing address.
```

File: reference/resources/address.yaml [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+    maxLength: 255
+    description: The City to use for this address.
+    example: "Greater London"
+    nullable: false
+
+  house_number_or_name:
+    type: string
+    minLength: 1
+    maxLength: 255
+    description: The house number or name to use for this address.
```
Comment (cbetta) [inline reference/resources/address.yaml:54]:
```suggestion
    description: |- 
      The house number or name for the billing address. Not all payment services use this field but some do.
```

File: reference/resources/address.yaml [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+    maxLength: 50
+    description: Postal code to use for this address.
+    example: "789123"
+    nullable: false
+
+  state:
+    type: string
+    minLength: 1
+    maxLength: 255
+    description: The City to use for this address.
```
Comment (cbetta) [inline reference/resources/address.yaml:46]:
```suggestion
    description: The city name for the billing address.
```

File: reference/resources/address.yaml [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+    maxLength: 2
+    description: Country to use for this address.
+    example: "GB"
+    nullable: false
+
+  postal_code:
+    type: string
+    minLength: 1
+    maxLength: 50
+    description: Postal code to use for this address.
```
Comment (cbetta) [inline reference/resources/address.yaml:38]:
```suggestion
    description: The postal code or zip code for the billing address.
```

File: reference/resources/address.yaml [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+    maxLength: 100
+    description: City to use for this address.
+    example: London
+    nullable: false
+
+  country:
+    type: string
+    minLength: 1
+    maxLength: 2
+    description: Country to use for this address.
```
Comment (cbetta) [inline reference/resources/address.yaml:30]:
```suggestion
    description: The country for the billing address.
```

File: reference/resources/address.yaml [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+  - state
+  - postal_code
+  - country
+
+properties:
+  city:
+    type: string
+    minLength: 1
+    maxLength: 100
+    description: City to use for this address.
```
Comment (cbetta) [inline reference/resources/address.yaml:22]:
```suggestion
    description: The city for the billing address.
```

File: reference/resources/address.yaml [hunk 28]
```diff
@@ -0,0 +1,80 @@
+---
+title: Address
+type: object
+description: |-
+  Billing details address to associate to a buyer.
```
Comment (cbetta) [inline reference/resources/address.yaml:5]:
```suggestion
  Billing address details to associate to a buyer.
```

**Review by cbetta — COMMENTED** [review]
Lots of copy changes requested. 

Additionally please run the linter locally and verify it doesn't throw any errors/

**Comment by alibeylan** [comment]
> Lots of copy changes requested.
> 
> Additionally please run the linter locally and verify it doesn't throw any errors/

@cbetta  Hmmm I did run the linter locally and no errors are found, will have another look though (:

File: reference/resources/billing_details.yaml [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+
+  last_name:
+    type: string
+    minLength: 1
+    maxLength: 255
+    example: Lunn
+    description: The last name, or family name, of the buyer.
+    nullable: true
+
+  email:
```
Comment (ipeluffo) [inline reference/resources/billing_details.yaml:27]:
```suggestion
  email_address:
```

File: reference/resources/tax_id.yaml [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+
+x-tags:
+  - Buyers
+
+required:
+  - kind
+  - value
+
+properties:
+  tax_id:
```
Comment (ipeluffo) [inline reference/resources/tax_id.yaml:15]:
```suggestion
  id:
```

File: reference/resources/tax_id.yaml [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+    type: string
+    minLength: 1
+    maxLength: 50
+    description: The tax ID for the buyer.
+    example: "12345678931"
+
+  tax_id_kind:
```
Comment (ipeluffo) [inline reference/resources/tax_id.yaml:22]:
```suggestion
  kind:
```

File: reference/resources/tax_id.yaml [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+    type: string
+    description: The unique name for this kind of tax ID.
```
Comment (ipeluffo) [inline reference/resources/tax_id.yaml:24]:
```suggestion
    description: The kind of tax ID.
```

**Review by cbetta — APPROVED** [review]

**Review by cbetta — APPROVED** [review]

**Pull request merged**